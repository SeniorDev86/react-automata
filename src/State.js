import React from 'react'
import PropTypes from 'prop-types'

class State extends React.Component {
  render() {
    return this.context.machineState === this.props.name
      ? this.props.children
      : null
  }
}

State.defaultProps = {
  children: null,
}

State.propTypes = {
  name: PropTypes.string.isRequired,
  children: PropTypes.node,
}

State.contextTypes = {
  machineState: PropTypes.string,
}

export default State
