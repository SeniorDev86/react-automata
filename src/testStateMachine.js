import React from 'react'
import TestRenderer from 'react-test-renderer'
import { withStateMachine } from './'

let visitedStates

const transition = (renderer, machineState, event, fixtures) => {
  const instance = renderer.getInstance()
  instance.setState({ machineState })

  if (event) {
    instance.handleTransition(event, fixtures)
  }

  return instance.state.machineState
}

const moveToNextState = (config, Component, machineState) => {
  visitedStates.push(machineState)

  machineState.split('.').reduce((states, state) => {
    const { on: events = {} } = states[state]

    Object.keys(events).forEach(event => {
      if (!visitedStates.includes(events[event])) {
        toMatchSnapshot(config, Component, machineState, event)
      }
    })

    return states[state].states
  }, config.machine.states)
}

const toMatchSnapshot = (config, Component, machineState, event) => {
  const initialData = config.fixtures ? config.fixtures.initialData : null
  const StateMachine = withStateMachine(config.machine, { initialData })(
    Component
  )
  const renderer = TestRenderer.create(<StateMachine />)
  const fixtures =
    config.fixtures && config.fixtures[machineState]
      ? config.fixtures[machineState][event]
      : null
  const nextMachineState = transition(renderer, machineState, event, fixtures)

  expect(renderer.toJSON()).toMatchSnapshot(nextMachineState)
  moveToNextState(config, Component, nextMachineState)
}

const testStateMachine = (config, Component) => {
  visitedStates = []
  toMatchSnapshot(config, Component, config.machine.initial)
}

export default testStateMachine
