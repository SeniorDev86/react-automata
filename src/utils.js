export const getComponentName = Component =>
  Component.displayName || Component.name || 'Component'

export const isStateless = Component =>
  !(Component.prototype && Component.prototype.isReactComponent)
