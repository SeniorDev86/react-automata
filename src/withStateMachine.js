import React from 'react'
import PropTypes from 'prop-types'
import { Machine } from 'xstate'

const getComponentName = Component =>
  Component.displayName || Component.name || 'Component'

const withStateMachine = (config, options = {}) => Component => {
  class StateMachine extends React.Component {
    machine = Machine(config)

    state = {
      action: null,
      componentState: null,
      machineState: this.machine.getInitialState(),
    }

    getChildContext() {
      return { machineState: this.state.machineState }
    }

    componentDidMount() {
      if (options.devTools && window.__REDUX_DEVTOOLS_EXTENSION__) {
        this.devTools = window.__REDUX_DEVTOOLS_EXTENSION__.connect({
          name: getComponentName(Component),
        })
        this.devTools.init({ machineState: this.state.machineState })

        this.unsubscribe = this.devTools.subscribe(message => {
          if (
            message.type === 'DISPATCH' &&
            message.payload.type === 'JUMP_TO_ACTION'
          ) {
            this.setState(JSON.parse(message.state))
          }
        })
      }
    }

    componentWillUnmount() {
      if (this.unsubscribe) {
        this.unsubscribe()
      }

      if (window.__REDUX_DEVTOOLS_EXTENSION__) {
        window.__REDUX_DEVTOOLS_EXTENSION__.disconnect()
      }
    }

    componentDidUpdate(prevProps, prevState) {
      if (
        prevState.machineState !== this.state.machineState &&
        this.state.action
      ) {
        if (this.instance.componentDidTransition) {
          this.instance.componentDidTransition(
            prevState.machineState,
            this.state.action,
            this.state.payload
          )
        }

        if (this.devTools) {
          this.devTools.send(this.state.action, {
            componentState: this.state.componentState,
            machineState: this.state.machineState,
          })
        }
      }
    }

    handleRef = element => {
      this.instance = element
    }

    handleTransition = (action, payload) => {
      if (this.instance.componentWillTransition) {
        this.instance.componentWillTransition(action, payload)
      }

      this.setState(prevState => ({
        action,
        componentState: { ...this.state.componentState, ...payload },
        machineState: this.machine
          .transition(prevState.machineState, action)
          .toString(),
        payload,
      }))
    }

    render() {
      return (
        <Component
          {...this.props}
          {...this.state.componentState}
          machineState={this.state.machineState}
          ref={this.handleRef}
          transition={this.handleTransition}
        />
      )
    }
  }

  StateMachine.childContextTypes = {
    machineState: PropTypes.string,
  }

  return StateMachine
}

export default withStateMachine
