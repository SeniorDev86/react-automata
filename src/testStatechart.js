import React from 'react'
import TestRenderer from 'react-test-renderer'
import { Machine } from 'xstate'
import { getShortestPaths } from 'xstate/lib/graph'
import idx from 'idx'
import invariant from 'invariant'
import { getContextValue } from './utils'
import withStatechart from './withStatechart'

const testStatechart = (config, Component) => {
  invariant(
    !Component.isStateMachine,
    `It seems you are testing a component wrapped into \`withStatechart\`, please use a base component instead.
    See https://github.com/MicheleBertoli/react-automata/issues/46`
  )

  const { channel, statechart } = config
  const machine = Machine(statechart)
  const paths = getShortestPaths(machine)

  Object.keys(paths).forEach(key => {
    const initialData = idx(config, _ => _.fixtures.initialData)
    const StateMachine = withStatechart(statechart, { channel })(Component)
    const renderer = TestRenderer.create(
      <StateMachine initialData={initialData} />
    )
    const instance = renderer.getInstance()

    paths[key].forEach(({ event, state }) => {
      const fixtures = idx(config, _ => _.fixtures[state][event])

      instance.handleTransition(event, fixtures)
    })

    const { machineState } = getContextValue(
      instance.getChildContext(),
      channel
    )

    expect(renderer.toJSON()).toMatchSnapshot(machineState)
  })
}

export default testStatechart
