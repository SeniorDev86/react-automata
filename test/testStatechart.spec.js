import React from 'react'
import { Action, State, testStatechart } from '../src'

const secondMachine = {
  initial: 'a',
  states: {
    a: {
      on: {
        SECOND_NEXT: 'b',
      },
      onEntry: 'enterBA',
    },
    b: {
      on: {
        SECOND_NEXT: 'a',
      },
      onEntry: 'enterBB',
    },
  },
}

const firstMachine = {
  initial: 'a',
  states: {
    a: {
      on: {
        FIRST_NEXT: 'b',
      },
    },
    b: {
      on: {
        FIRST_NEXT: 'a',
      },
      onEntry: 'enterB',
      ...secondMachine,
    },
  },
}

test('action', () => {
  const App = () => (
    <div>
      <Action initial hide="enterB">
        a
      </Action>
      <Action show="enterBA">b.a</Action>
      <Action show="enterBB">b.b</Action>
    </div>
  )

  testStatechart({ statechart: firstMachine }, App)
})

test('state', () => {
  const App = () => (
    <div>
      <State value="a">a</State>
      <State value="b.a">b.a</State>
      <State value="b.b">b.b</State>
    </div>
  )

  testStatechart({ statechart: firstMachine }, App)
})
