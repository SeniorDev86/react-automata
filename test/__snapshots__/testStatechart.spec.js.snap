// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a 1`] = `
<div>
  a
</div>
`;

exports[`a 2`] = `
<div>
  a
</div>
`;

exports[`b.a 1`] = `
<div>
  b.a
</div>
`;

exports[`b.a 2`] = `
<div>
  b.a
</div>
`;

exports[`b.b 1`] = `
<div>
  b.b
</div>
`;

exports[`b.b 2`] = `
<div>
  b.b
</div>
`;
